File: backend/internal/cache/cache.go
// Package cache is the TTL cache with in-flight dedup, ported from
// src/lib/server/cache.ts. Like the original it is a process-wide singleton:
// the point is cross-panel sharing (the metrics panel peeks the kube panel,
// the incident report peeks everything), and passing one instance through
// every constructor would buy nothing but plumbing.
package cache

import (
	"sync"
	"time"
)

type entry struct {
	at   time.Time
	ttl  time.Duration
	data any
	err  error
}

var (
	mu       sync.Mutex
	store    = map[string]*entry{}
	inflight = map[string]*call{}
)

type call struct {
	done chan struct{}
	data any
	err  error
}

// Cached returns the fresh value under key, or runs fn once — concurrent
// callers for the same key share one upstream call. With failTTL > 0 a failure
// is cached briefly too, so a broken upstream is not hammered on every poll.
func Cached[T any](key string, ttl time.Duration, fn func() (T, error), failTTL time.Duration) (T, error) {
	mu.Lock()
	if e, ok := store[key]; ok && time.Since(e.at) < e.ttl {
		mu.Unlock()
		if e.err != nil {
			var zero T
			return zero, e.err
		}
		return e.data.(T), nil
	}
	if c, ok := inflight[key]; ok {
		mu.Unlock()
		<-c.done
		if c.err != nil {
			var zero T
			return zero, c.err
		}
		return c.data.(T), nil
	}
	c := &call{done: make(chan struct{})}
	inflight[key] = c
	mu.Unlock()

	data, err := fn()
	c.data, c.err = data, err

	mu.Lock()
	if err == nil {
		store[key] = &entry{at: time.Now(), ttl: ttl, data: data}
	} else if failTTL > 0 {
		store[key] = &entry{at: time.Now(), ttl: failTTL, err: err}
	}
	delete(inflight, key)
	mu.Unlock()
	close(c.done)
	return data, err
}

// Peek returns the cached value even when stale, matching the TS peekCached.
func Peek[T any](key string) (T, bool) {
	mu.Lock()
	defer mu.Unlock()
	if e, ok := store[key]; ok && e.err == nil {
		if v, ok := e.data.(T); ok {
			return v, true
		}
	}
	var zero T
	return zero, false
}

func Put(key string, ttl time.Duration, data any) {
	mu.Lock()
	defer mu.Unlock()
	store[key] = &entry{at: time.Now(), ttl: ttl, data: data}
}

// Invalidate drops every key with the prefix; "" clears the lot.
func Invalidate(prefix string) {
	mu.Lock()
	defer mu.Unlock()
	for k := range store {
		if len(k) >= len(prefix) && k[:len(prefix)] == prefix {
			delete(store, k)
		}
	}
}

File: backend/cache/cache.ts
// TTL cache with in-flight dedup.
//
// A process-wide singleton on purpose: the point is cross-panel sharing (the
// metrics panel peeks the kube panel, the incident report peeks everything),
// and passing one instance through every constructor would buy nothing but
// plumbing.

interface Entry {
  at: number;
  ttl: number;
  data?: unknown;
  err?: Error;
}

const store = new Map<string, Entry>();
const inflight = new Map<string, Promise<unknown>>();

/**
 * Returns the fresh value under key, or runs fn once — concurrent callers for
 * the same key share one upstream call. With failTtl > 0 a failure is cached
 * briefly too, so a broken upstream is not hammered on every poll.
 */
export async function cached<T>(
  key: string,
  ttl: number,
  fn: () => Promise<T>,
  failTtl = 0,
): Promise<T> {
  const hit = store.get(key);
  if (hit && Date.now() - hit.at < hit.ttl) {
    if (hit.err) throw hit.err;
    return hit.data as T;
  }

  const running = inflight.get(key);
  if (running) return running as Promise<T>;

  const promise = (async () => {
    try {
      const data = await fn();
      store.set(key, { at: Date.now(), ttl, data });
      return data;
    } catch (e) {
      const err = e instanceof Error ? e : new Error(String(e));
      if (failTtl > 0) store.set(key, { at: Date.now(), ttl: failTtl, err });
      throw err;
    } finally {
      inflight.delete(key);
    }
  })();

  inflight.set(key, promise);
  return promise;
}

/** Returns the cached value even when stale. */
export function peek<T>(key: string): T | undefined {
  const e = store.get(key);
  if (e && !e.err) return e.data as T;
  return undefined;
}

export function put(key: string, ttl: number, data: unknown): void {
  store.set(key, { at: Date.now(), ttl, data });
}

/** Drops every key with the prefix; "" clears the lot. */
export function invalidate(prefix: string): void {
  for (const k of store.keys()) {
    if (k.startsWith(prefix)) store.delete(k);
  }
}
